safemysql.class.php

```diff
@@ -488,14 +488,34 @@ protected function rawQuery($query)
 
 	protected function prepareQuery($args)
 	{
-		$query = '';
-		$raw   = array_shift($args);
-		$array = preg_split('~(\?[nsiuap])~u',$raw,null,PREG_SPLIT_DELIM_CAPTURE);
+		$query  = '';
+		$raw    = array_shift($args);
+		$values = $args;
+
+		$namedPlaceholders = false;
+		if (preg_match('~(\?[nsiuap](?:\:[a-zA-Z0-9_]+))~u', $raw))
+		{
+			$regexp = '~(\?[nsiuap](?:\:[a-zA-Z0-9_]+)?)~u';
+			$namedPlaceholders = true;
+		} else {
+			$regexp = '~(\?[nsiuap])~u';
+		}
+
+		$array = preg_split('~(\?[nsiuap](?:\:[a-zA-Z0-9_]+)?)~u',$raw,null,PREG_SPLIT_DELIM_CAPTURE);
+
+		if ($namedPlaceholders)
+		{
+			$args = $args[0];
+			$values = array_filter($args, function($key) {
+				return is_int($key);
+			}, ARRAY_FILTER_USE_KEY);
+		}
+
 		$anum  = count($args);
 		$pnum  = floor(count($array) / 2);
-		if ( $pnum != $anum )
+		if ( $pnum > $anum )
```

Comment:
I think it might be better to check it in case `$namedPlaceholders` is `false` otherwise only check array key presence. Also in case of using named placeholders we have ability to reuse placeholders, so in this case `$anum` is greater than `$pnum`.
I.e.
```
$row = $db->getRow('SELECT * FROM `table` WHERE `username` = ?s:user OR `email` = ?s:user', array('user' => $_POST['user']))
```